map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp

```diff
@@ -49,21 +49,62 @@
 #include <lanelet2_io/Io.h>
 #include <lanelet2_projection/UTM.h>
 
+#include <filesystem>
 #include <string>
 
-Lanelet2MapLoaderNode::Lanelet2MapLoaderNode(const rclcpp::NodeOptions & options)
-: Node("lanelet2_map_loader", options)
+namespace
 {
-  const auto adaptor = component_interface_utils::NodeAdaptor(this);
+bool isOsmFile(const std::string & f)
+{
+  if (std::filesystem::is_directory(f)) {
+    return false;
+  }
 
-  // subscription
-  adaptor.init_sub(
-    sub_map_projector_info_,
-    [this](const MapProjectorInfo::Message::ConstSharedPtr msg) { on_map_projector_info(msg); });
+  const std::string ext = std::filesystem::path(f).extension();
 
+  if (ext != ".osm" && ext != ".OSM") {
+    return false;
+  }
+
+  return true;
+}
+}  // namespace
+
+Lanelet2MapLoaderNode::Lanelet2MapLoaderNode(const rclcpp::NodeOptions & options)
+: Node("lanelet2_map_loader", options)
+{
   declare_parameter<std::string>("lanelet2_map_path");
   declare_parameter<double>("center_line_resolution");
   declare_parameter<bool>("use_waypoints");
+
+  if (isOsmFile(get_parameter("lanelet2_map_path").as_string())) {
```

Comment:
I don't want the implementation to switch based on the number of files.
Can we generate dummy metadata when there is only one file, [similar to how it's done for PCD](https://github.com/autowarefoundation/autoware.universe/blob/f65af78d7515c44f1b1b09cd0c0777852e6894ac/map/map_loader/src/pointcloud_map_loader/pointcloud_map_loader_node.cpp#L106-L120)?

Reply:
We can create a dummy metadata file as in the point cloud loading, but my concern is:
- If we do this, we create a differential map loader object every time. This applies whether we have a single OSM map or multiple OSM maps. As a result we see a lot of useless topic and service when we use single osm file.
- We should remove the map loading functions ([link 1](https://github.com/autowarefoundation/autoware.universe/blob/f65af78d7515c44f1b1b09cd0c0777852e6894ac/map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp#L135), [link 2](https://github.com/autowarefoundation/autoware.universe/blob/f65af78d7515c44f1b1b09cd0c0777852e6894ac/map/map_loader/src/lanelet2_map_loader/lanelet2_map_loader_node.cpp#L185)).
- I am not sure how we can calculate the origin positions of an OSM file without reading the OSM file. Additionally, if we can somehow read the origins, we need to project the latitude and longitude from the OSM file. We also need projection information and a projector inside the get_lanelet2_metadata function, but both are missing in the current implementation.

So, can you inform me how we can do this in the best way? I cannot find a good solution.

Reply:
If <arg name="enable_differential_map_loading" default="true"> is set in the launch file, it will not work correctly for a single lanelet2 file. Therefore, even for a single lanelet2 file, if `enable_differential_map_loading` is true, we should create a differential map loader object.

If it is hard to implement this, please make it a runtime error when `enable_differential_map_loading` is true for a single lanelet2 file (when the metadata file does not exist).
But if we read the OSM file, I believe we can calculate the origin, and if we read the MapProjectorInfo, we can project the latitude and longitude.

Reply:
For now I just changed following two as you mentioned:
- remove the if statement with `if (isOsmFile(get_parameter("lanelet2_map_path").as_string())` and put `enable_differential_map_loading` check.
- runtime error when enable_differential_map_loading is true for a single lanelet2 file

Also I add a TODO for dummy metadata for single lanelet2 file and I want to ask something. If we want to use MapProjectorInfo, we should read metadata in `on_map_projector_info`. Do you think it would be okay for us to do this?
If yes I will update the code.

One more question: If we remove the if statement from constructor of Lanelet2MapLoaderNode, is it okay for you?

Reply:
> If we want to use MapProjectorInfo, we should read metadata in on_map_projector_info. Do you think it would be okay for us to do this?

Yes, I think it's okay.


>  If we remove the if statement from constructor of Lanelet2MapLoaderNode, is it okay for you?

Yes, but I think the if statement will remain.
Here's the code I envision:
```
Lanelet2MapLoaderNode::Lanelet2MapLoaderNode(const rclcpp::NodeOptions & options) {
  ...
  const auto adaptor = component_interface_utils::NodeAdaptor(this);
  adaptor.init_sub(
    sub_map_projector_info_,
    [this](const MapProjectorInfo::Message::ConstSharedPtr msg) { on_map_projector_info(msg); });

  if (get_parameter("enable_differential_map_loading").as_bool()) {
    differential_loader_module_ = std::make_unique<Lanelet2DifferentialLoaderModule>(
      this, /*lanelet2_metadata_dict, x_resolution, y_resolution,*/
      get_parameter("center_line_resolution").as_double());
  }
}

void Lanelet2MapLoaderNode::on_map_projector_info(const MapProjectorInfo::Message::ConstSharedPtr msg) {
  ...
  lanelet2_map = load_lanelet2_file(path);
  lanelet2_map = projection(lanelet2_map, msg);
  publish(lanelet2_map);

  if (enable_differential_map_loading) {
    if (exist_medata_file) {
      metadata = load_metadata_file(path);
    }
    else {
      if (single_lanelet2_map) {
        metadata = create_dummy_metadata(lanelet2_map);
        metadata = projection(metadata, msg);
      }
      else {
       runtime_error("metadata_file_is_not_found");
      }
    differential_loader_module_.set_metadata(metadata);
    publish(metadata);
    }

}
```

Reply:
I followed the steps listed below according to the pseudo code you sent:

- I couldn't read map first like you did because when `enable_differential_map_loading` is `true` and the map is divided, it will return an error because it will try to load folder path.

- I added a `setProjectionInfo` function to `Lanelet2DifferentialLoaderModule` instead of `map_interface::MapProjectorInfo` subscriber which located in constructor of `Lanelet2DifferentialLoaderModule` because `Lanelet2MapLoaderNode::on_map_projector_info` function blocks the subscriber and also `Lanelet2MapLoaderNode::on_map_projector_info` function needs to use `Lanelet2DifferentialLoaderModule`. If we want to use `map_interface::MapProjectorInfo` subscriber in both class, I guess, we have to make `map_projection_loader` to multi thread executor. 

- For dummy_metadata I used dummy x and y offsets too because in normal we read offsets from `lanelet_metadata_file`.

- `differentialLanelet2Load` function updated and it is a public function now and we are calling it in `Lanelet2MapLoaderNode` to load entire map for once using all tiles from metadata when `enable_differential_map_loading` is `true` whether metadata is available or not (if it not available use dummy metadata).

- In `lanelet2_map_loader_node.cpp` on line 169, I changed the value of lanelet2_filename to the first tile path using metadata. This is because in the following parts of the code, lanelet2_filename is used to get the version from the osm file. If we are using divided OSM files, lanelet2_filename would be a folder path, so we need to modify it accordingly. I also added a TODO there to improve this approach in the future.

The result is not exactly the same as the pseudo code you prepared. Could you check it when you are available?

Reply:
@StepTurtle Great! Thank you so much!
I have one question.
In the `get_dummy_lanelet2_metadata` function,
what are `map->pointLayer.begin()->x()`  and `map->pointLayer.begin()->y()`?
I'm particularly unsure about the reason for dividing by 3.
```
  tile.min_x = map->pointLayer.begin()->x() - (map->pointLayer.begin()->x() / 3);
  tile.min_y = map->pointLayer.begin()->y() - (map->pointLayer.begin()->y() / 3);
  x_resolution = (map->pointLayer.begin()->x() * 2) / 3;
  y_resolution = (map->pointLayer.begin()->y() * 2) / 3;
```

Reply:
Hi,

It would have made sense to add an explanation for this calculations, I forgot, sorry.

`map->pointLayer.begin()->x()` and `map->pointLayer.begin()->y()` are just the first node from `osm` file (They can be any point within the map, they are not ordered data).

![Screenshot from 2024-08-22 13-20-54](https://github.com/user-attachments/assets/e69614be-e65f-4a3a-aa5a-344a7c531314)


I am trying to calculate the precise boundaries for a lanelet map. The calculations create boundaries similar to the diagram below. However, the method I used is somewhat simplistic because I assumed that finding the smallest and largest points of the lanelet map could save a lot of time. Maybe we think to change that. Do you have any idea on it?
 
![Untitled Diagram drawio(1)](https://github.com/user-attachments/assets/aec5491d-c5eb-4d1c-8a8d-d38e32fd7d89)

Reply:
@StepTurtle Hi.
How does [lanelet2-map-tile-generator](https://github.com/leo-drive/lanelet2-map-tile-generator/tree/master) calculate min_x, min_y, and resolution? Would it be difficult to calculate them in the same way? Or are you concerned about the processing time?

Reply:
If it's difficult to calculate, since this is just a dummy, I think it's fine to input arbitrarily large values, like:

min_x: 0.0
min_y: 0.0
x_resolution: 100000.0
y_resolution: 100000.0

It would be even better if these dummy values could be overwritten in a parameter file.

Reply:
The lanelet-tile-generator takes the MGRS code and tile ranges as input. Using these two, it divides the MGRS into equal parts, and as a result, the minimum and maximum values are determined based on the corners of the equal parts, which might not be suitable for our current situation.

I changed the values with `100000`. I guess the `100000.0` is max value for x and y in single MGRS grid. Than it should load map for all possible initial positions.

> It would be even better if these dummy values could be overwritten in a parameter file.

I have a slight concern about this. If we save the `lanelet2_map_metadata.yaml` file inside the map folder, I'm not sure how we should fill in the osm file name. Normally, it would be `X.osm`, and there would be a file named `.../lanelet2_map.osm/X.osm`, but in our case, this might cause an error.

Reply:
@StepTurtle Hi
I was thinking of defining the parameters in the [lanelet2_map_loader.param.yaml](https://github.com/autowarefoundation/autoware.universe/blob/main/map/map_loader/config/lanelet2_map_loader.param.yaml) ([also in the autoware.launch](https://github.com/autowarefoundation/autoware_launch/blob/main/autoware_launch/config/map/lanelet2_map_loader.param.yaml))

```
/**:
  ros__parameters:
    allow_unsupported_version: true             # flag to load unsupported format_version anyway. If true, just prints warning.
    center_line_resolution: 5.0                 # [m]
    use_waypoints: true                         # "centerline" in the Lanelet2 map will be used as a "waypoints" tag.
    lanelet2_map_path: $(var lanelet2_map_path) # The lanelet2 map path
    dummy_metadata:                             # When the dynamic map is true and the map is a single file, this parameter will generate the metadata.
      min_x: 0.0
      min_y: 0.0
      x_resolution: 100000.0
      y_resolution: 100000.0
```

Reply:
Sorry @YamatoAndo, I misunderstand you.

 I add the parameters you mentioned to both config file.

Reply:
@StepTurtle Thank you!
And cloud you please also modify the [lanelet2_map_loader.schema.json](https://github.com/autowarefoundation/autoware.universe/blob/main/map/map_loader/schema/lanelet2_map_loader.schema.json) ?

Reply:
OK, updated.

Reminder, I did not put the new parameters to [required list](https://github.com/autowarefoundation/autoware.universe/blob/fe953a40cd4a2c09a7cd82d6f25c1d09b2a9d86c/map/map_loader/schema/lanelet2_map_loader.schema.json#L30) because of these looks not required for default usages. Let me know if you want to update this list too.

Reply:
While reviewing the code, I realized that a situation had been overlooked. When using dummy metadata, `dynamic_map_loading_map_radius_` may be incompatible with the limits we provide. Therefore, if a user sets `enable_differential_map_loading` to `true` and provides a single map, the map may not be found with the default `dynamic_map_loading_map_radius_` value. To address this, I added a check to determine whether the current position is within the map. It looks OK right now.

related commit: https://github.com/autowarefoundation/autoware.universe/pull/6241/commits/30523b6099fe765a9630a89d607337f659a11c07